File: backend/models/share.go
package models

import (
	"time"
)

// ResourceType represents the type of resource being shared
type ResourceType string

const (
	ResourceTypeDashboard ResourceType = "dashboard"
	ResourceTypeQuery     ResourceType = "query"
)

// SharePermission represents the permission level for a share
type SharePermission string

const (
	SharePermissionView  SharePermission = "view"
	SharePermissionEdit  SharePermission = "edit"
	SharePermissionAdmin SharePermission = "admin"
)

// ShareStatus represents the status of a share
type ShareStatus string

const (
	ShareStatusActive  ShareStatus = "active"
	ShareStatusRevoked ShareStatus = "revoked"
	ShareStatusExpired ShareStatus = "expired"
	ShareStatusPending ShareStatus = "pending"
)

// Share represents a resource share between users
type Share struct {
	ID           string          `gorm:"type:uuid;primaryKey;default:gen_random_uuid()" json:"id"`
	ResourceType ResourceType    `gorm:"type:varchar(50);not null;index" json:"resource_type"`
	ResourceID   string          `gorm:"type:text;not null;index" json:"resource_id"`
	SharedBy     string          `gorm:"type:text;not null;index" json:"shared_by"`
	SharedWith   *string         `gorm:"type:text;index" json:"shared_with,omitempty"`  // User ID for internal shares
	SharedEmail  *string         `gorm:"type:text;index" json:"shared_email,omitempty"` // Email for external invites
	Permission   SharePermission `gorm:"type:varchar(20);not null;default:'view'" json:"permission"`
	PasswordHash *string         `gorm:"type:text" json:"-"` // Never return password hash
	ExpiresAt    *time.Time      `gorm:"type:timestamp" json:"expires_at,omitempty"`
	Status       ShareStatus     `gorm:"type:varchar(20);not null;default:'active';index" json:"status"`
	AcceptedAt   *time.Time      `gorm:"type:timestamp" json:"accepted_at,omitempty"`
	Message      *string         `gorm:"type:text" json:"message,omitempty"`
	CreatedAt    time.Time       `gorm:"autoCreateTime" json:"created_at"`
	UpdatedAt    time.Time       `gorm:"autoUpdateTime" json:"updated_at"`

	// Relationships
	SharedByUser   User  `gorm:"foreignKey:SharedBy;references:ID" json:"shared_by_user,omitempty"`
	SharedWithUser *User `gorm:"foreignKey:SharedWith;references:ID" json:"shared_with_user,omitempty"`
}

// TableName specifies the table name for Share
func (Share) TableName() string {
	return "shares"
}

// IsExpired checks if the share has expired
func (s *Share) IsExpired() bool {
	if s.ExpiresAt == nil {
		return false
	}
	return time.Now().After(*s.ExpiresAt)
}

// IsActive checks if the share is active and not expired
func (s *Share) IsActive() bool {
	if s.Status != ShareStatusActive {
		return false
	}
	return !s.IsExpired()
}

// RequiresPassword checks if the share requires a password
func (s *Share) RequiresPassword() bool {
	return s.PasswordHash != nil && *s.PasswordHash != ""
}

// HasEditPermission checks if the share grants edit permission
func (s *Share) HasEditPermission() bool {
	return s.Permission == SharePermissionEdit || s.Permission == SharePermissionAdmin
}

// HasAdminPermission checks if the share grants admin permission
func (s *Share) HasAdminPermission() bool {
	return s.Permission == SharePermissionAdmin
}

// ShareFilter represents filter criteria for querying shares
type ShareFilter struct {
	ResourceType   *ResourceType    `json:"resource_type,omitempty"`
	ResourceID     *string          `json:"resource_id,omitempty"`
	SharedBy       *string          `json:"shared_by,omitempty"`
	SharedWith     *string          `json:"shared_with,omitempty"`
	SharedEmail    *string          `json:"shared_email,omitempty"`
	Status         *ShareStatus     `json:"status,omitempty"`
	Permission     *SharePermission `json:"permission,omitempty"`
	IncludeExpired bool             `json:"include_expired,omitempty"`
	Limit          int              `json:"limit,omitempty"`
	Offset         int              `json:"offset,omitempty"`
}

// ShareCreateRequest represents the request to create a share
type ShareCreateRequest struct {
	ResourceType ResourceType    `json:"resource_type" validate:"required,oneof=dashboard query"`
	ResourceID   string          `json:"resource_id" validate:"required"`
	SharedWith   *string         `json:"shared_with,omitempty"`  // User ID
	SharedEmail  *string         `json:"shared_email,omitempty"` // Email for external invite
	Permission   SharePermission `json:"permission" validate:"required,oneof=view edit admin"`
	Password     *string         `json:"password,omitempty"`
	ExpiresAt    *time.Time      `json:"expires_at,omitempty"`
	Message      *string         `json:"message,omitempty"`
}

// ShareUpdateRequest represents the request to update a share
type ShareUpdateRequest struct {
	Permission *SharePermission `json:"permission,omitempty" validate:"omitempty,oneof=view edit admin"`
	Password   *string          `json:"password,omitempty"`
	ExpiresAt  *time.Time       `json:"expires_at,omitempty"`
	Message    *string          `json:"message,omitempty"`
}

// ShareAccessCheck represents the result of checking share access
type ShareAccessCheck struct {
	HasAccess        bool            `json:"has_access"`
	Permission       SharePermission `json:"permission,omitempty"`
	ShareID          string          `json:"share_id,omitempty"`
	RequiresPassword bool            `json:"requires_password"`
}

// ShareWithDetails represents a share with user details
type ShareWithDetails struct {
	Share
	ResourceName string `json:"resource_name,omitempty"`
}

// ValidateSharePermission validates a share permission string
func ValidateSharePermission(permission string) (SharePermission, bool) {
	switch SharePermission(permission) {
	case SharePermissionView, SharePermissionEdit, SharePermissionAdmin:
		return SharePermission(permission), true
	default:
		return "", false
	}
}

// ValidateResourceType validates a resource type string
func ValidateResourceType(resourceType string) (ResourceType, bool) {
	switch ResourceType(resourceType) {
	case ResourceTypeDashboard, ResourceTypeQuery:
		return ResourceType(resourceType), true
	default:
		return "", false
	}
}

File: frontend/types/share.ts
/**
 * Share and Embed Token Type Definitions
 * 
 * TypeScript interfaces for Advanced Sharing feature (TASK-088 to TASK-091)
 * Matches backend models in backend/models/share.go and backend/models/embed_token.go
 */

// ============================================================================
// Core Entity Types
// ============================================================================

/**
 * Resource types that can be shared
 */
export type ResourceType = 'dashboard' | 'query'

/**
 * Permission levels for shares
 */
export type SharePermission = 'view' | 'edit' | 'admin'

/**
 * Share statuses
 */
export type ShareStatus = 'active' | 'revoked' | 'expired' | 'pending'

/**
 * User information for shares
 */
export interface ShareUser {
    id: string
    username: string
    email: string
    name?: string
}

/**
 * Share entity representing a resource share
 */
export interface Share {
    id: string
    resource_type: ResourceType
    resource_id: string
    shared_by: string
    shared_with?: string
    shared_email?: string
    permission: SharePermission
    expires_at?: string
    status: ShareStatus
    accepted_at?: string
    message?: string
    created_at: string
    updated_at: string

    // Relationships (populated when fetching)
    shared_by_user?: ShareUser
    shared_with_user?: ShareUser
}

/**
 * Share with resource details
 */
export interface ShareWithDetails extends Share {
    resource_name?: string
}

// ============================================================================
// Share API Request Types
// ============================================================================

/**
 * Request to create a new share
 */
export interface CreateShareRequest {
    resource_type: ResourceType
    resource_id: string
    shared_with?: string
    shared_email?: string
    permission: SharePermission
    password?: string
    expires_at?: string
    message?: string
}

/**
 * Request to update a share
 */
export interface UpdateShareRequest {
    permission?: SharePermission
    password?: string
    expires_at?: string
    message?: string
}

/**
 * Request to accept a pending share
 */
export interface AcceptShareRequest {
    password?: string
}

/**
 * Request to validate share access with password
 */
export interface ValidateShareAccessRequest {
    share_id: string
    password?: string
}

// ============================================================================
// Share API Response Types
// ============================================================================

/**
 * Response for share access check
 */
export interface ShareAccessCheck {
    has_access: boolean
    permission?: SharePermission
    share_id?: string
    requires_password: boolean
}

/**
 * Response for validating share access
 */
export interface ValidateShareAccessResponse {
    valid: boolean
    share_id: string
    resource_type: ResourceType
    resource_id: string
    permission: SharePermission
}

/**
 * Filter for querying shares
 */
export interface ShareFilter {
    resource_type?: ResourceType
    status?: ShareStatus
    include_expired?: boolean
}

/**
 * Response for list of shares
 */
export interface GetSharesResponse {
    shares: Share[]
    total: number
}

// ============================================================================
// Embed Token Types
// ============================================================================

/**
 * Embed token entity for external embedding
 */
export interface EmbedToken {
    id: string
    resource_type: ResourceType
    resource_id: string
    token: string
    created_by: string
    allowed_domains: string[]
    allowed_ips: string[]
    expires_at?: string
    view_count: number
    last_viewed_at?: string
    is_revoked: boolean
    revoked_at?: string
    revoked_by?: string
    description?: string
    created_at: string
    updated_at: string

    // Relationships
    creator?: ShareUser
}

/**
 * Embed token with resource statistics
 */
export interface EmbedTokenWithStats extends EmbedToken {
    resource_name?: string
}

// ============================================================================
// Embed Token API Request Types
// ============================================================================

/**
 * Request to create a new embed token
 */
export interface CreateEmbedTokenRequest {
    resource_type: ResourceType
    resource_id: string
    allowed_domains?: string[]
    allowed_ips?: string[]
    expires_at?: string
    description?: string
}

/**
 * Request to update an embed token
 */
export interface UpdateEmbedTokenRequest {
    allowed_domains?: string[]
    allowed_ips?: string[]
    expires_at?: string
    description?: string
}

// ============================================================================
// Embed Token API Response Types
// ============================================================================

/**
 * Response for embed token validation
 */
export interface EmbedTokenValidationResult {
    is_valid: boolean
    token_id?: string
    resource_type?: ResourceType
    resource_id?: string
    error?: string
}

/**
 * Response for validating embed token
 */
export interface ValidateEmbedTokenResponse {
    valid: boolean
    token_id: string
    resource_type: ResourceType
    resource_id: string
    view_count: number
}

/**
 * Filter for querying embed tokens
 */
export interface EmbedTokenFilter {
    resource_type?: ResourceType
    resource_id?: string
    include_expired?: boolean
    include_revoked?: boolean
}

/**
 * Response for list of embed tokens
 */
export interface GetEmbedTokensResponse {
    tokens: EmbedToken[]
    total: number
}

// ============================================================================
// UI State Types
// ============================================================================

/**
 * Share dialog state
 */
export interface ShareDialogState {
    isOpen: boolean
    resourceType: ResourceType
    resourceId: string
    resourceName: string
    shares: Share[]
    isLoading: boolean
    error: string | null
}

/**
 * Embed dialog state
 */
export interface EmbedDialogState {
    isOpen: boolean
    resourceType: ResourceType
    resourceId: string
    resourceName: string
    tokens: EmbedToken[]
    isLoading: boolean
    error: string | null
}

/**
 * New share form state
 */
export interface NewShareFormState {
    recipientType: 'user' | 'email'
    recipient: string
    permission: SharePermission
    requirePassword: boolean
    password: string
    setExpiration: boolean
    expirationDate: string
    message: string
}

/**
 * New embed token form state
 */
export interface NewEmbedTokenFormState {
    allowedDomains: string
    allowedIPs: string
    setExpiration: boolean
    expirationDate: string
    description: string
}

/**
 * Permission option for UI
 */
export interface PermissionOption {
    value: SharePermission
    label: string
    description: string
    icon: string
}
